### PR TITLE
KAFKA-15131: Improve RemoteStorageManager exception handling documentation

### PR DESCRIPTION
**Pull request by jeqo**
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteStorageManager.java
Comment (divijvaidya) [inline storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteStorageManager.java]: Perhaps, we can be more descriptive here. My suggestion is to add the following (free free to change the wording):

`e.g. Transaction index may not exist because segments create prior to version 2.8.0 will not have transaction index associated with them.`

`@throws RemoteResourceNotFoundException the requested index is not found in the remote storage. The caller of this function are encouraged to re-create the indexes from the segment as the suggested way of handling this error.`

**Review by divijvaidya — COMMENTED** [review]
The proposed clarification makes sense to me. Added some more details to your description. We should also modify the KIP with this.

@showuon @satishd please take a look.

**Comment by showuon** [comment]
I'll take a look this week.

File: storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteStorageManager.java
Comment (jeqo) [reply storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteStorageManager.java]: thanks @divijvaidya, applying this suggestion

**Comment by jeqo** [comment]
@divijvaidya adding the following changes to the KIP: https://cwiki.apache.org/confluence/pages/diffpagesbyversion.action?pageId=97554472&selectedPageVersions=363&selectedPageVersions=362

**Comment by divijvaidya** [comment]
Thanks @jeqo. Looks good to me but we will wait for Luke and/or Satish to look into this before we merge this one in.

**Comment by jeqo** [comment]
Sure, make sense

File: storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteStorageManager.java
Comment (showuon) [inline storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteStorageManager.java]: I think this line should be put below:

`* @throws RemoteResourceNotFoundException the requested index is not found in the remote storage (e.g. Transaction index may not exist because segments create prior to version 2.8.0 will not have transaction index associated with them.).`

WDYT?

File: storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteStorageManager.java
Comment (showuon) [inline storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteStorageManager.java]: This is confused because transactionIndex should be expected to be non-existed. Maybe we can put it this way:
`The caller of this function are encouraged to re-create the indexes from the segment as the suggested way of handling this error if the index is expected to be existed.`

WDYT?

**Review by showuon — COMMENTED** [review]
@jeqo , thanks for the PR. Left some comments.

**Comment by jeqo** [comment]
@showuon I've rephrased the comments, lmk how it looks now. Thanks!

**Review by showuon — APPROVED** [review]
LGTM! Thanks for the patch!

File: storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteStorageManager.java
Comment (satishd) [inline storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteStorageManager.java:79]: Newly added statement gives more clarity on idempotency. 

The unique id sent as part of `RemoteLogSegmentMetadata` can be used to avoid overwrites in RSM implementation. Caller makes sure it sends unique id for multiple invocations.

File: storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteStorageManager.java
Comment (satishd) [inline storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteStorageManager.java]: minor typo: `segments create prior to version` -> `segments created prior to version`

**Review by satishd — COMMENTED** [review]
Thanks @jeqo for the PR, left a couple of minor comments.

**Review by satishd — APPROVED** [review]
Thanks @jeqo for addressing the comments. LGTM

**Review by divijvaidya — APPROVED** [review]

**Pull request merged**
